### PR TITLE
Update coverity violation for SMP

### PR DESCRIPTION
**Pull request by chinglee-iot**
Fix coverity violations for SMP diff part.

Description
-----------

Fix coverity violations for SMP diff part.
SMP diff part is generated by comparing the coverity report of **single core** and **SMP**.

The coverity violations with SMP are already fixed in the implementation. The coverify report only shows violations in single core ( common code ). Single core FreeRTOS are checked by another tool "Lint". The reason for violation is documented with lint suppression comment.

The following is a lint suppression example:
```C
/*lint !e9079 All values returned by pvPortMalloc() have at least the alignment required 
 *by the MCU's stack and this allocation allocates a struct that has the alignment requirements of a pointer. */
```

The tasks.c coverity errors for single core

Checker | Violations | Category
-- | -- | --
MISRA C-2012 Directive 4.6 | 7 | Advisory
MISRA C-2012 Rule 10.1 | 3 | Required
MISRA C-2012 Rule 10.3 | 4 | Required
MISRA C-2012 Rule 10.4 | 51 | Required
MISRA C-2012 Rule 11.3 | 3 | Required
MISRA C-2012 Rule 11.4 | 2 | Advisory
MISRA C-2012 Rule 13.2 | 2 | Required
MISRA C-2012 Rule 14.4 | 37 | Required
MISRA C-2012 Rule 15.4 | 1 | Advisory
MISRA C-2012 Rule 17.7 | 4 | Required
MISRA C-2012 Rule 17.8 | 10 | Advisory
MISRA C-2012 Rule 20.5 | 1 | Advisory
MISRA C-2012 Rule 21.6 | 3 | Required
MISRA C-2012 Rule 5.9 | 2 | Advisory
MISRA C-2012 Rule 8.13 | 18 | Advisory
MISRA C-2012 Rule 8.4 | 2 | Required
MISRA C-2012 Rule 8.9 | 7 | Advisory
MISRA C-2012 Rule 9.1 | 1 | Mandatory

</body>
</html>

Test Steps
-----------

**Generate SMP coverity report**
1. Clone the SMP integration [FreeRTOS repo](https://github.com/chinglee-iot/FreeRTOS/tree/smp-integration-coverity-test).
2. cd FreeRTOS/Test/Coverity
3. Update the compiler configuration in Coverity
`cov-configure --force --compiler cc --comptype gcc -- -std=gnu90`
4. Build the (pseudo) application
`cov-build --emit-complementary-info --dir cov-out make multicore1`
or
`cov-build --emit-complementary-info --dir cov-out make multicore2`
5. Go to the Coverity output directory (cov-out) and begin Coverity static analysis
`cd cov-out/`
`cov-analyze --dir . --coding-standard-config ../coverity_misra.config`
6. Generate the report with the following command
`cov-format-errors --dir . --file "Source/tasks.c" --exclude-files '(/Test/|/Source/portable/)' --html-output html-out;`

**Generate single core coverity report**
1. Clone the SMP integration [FreeRTOS repo](https://github.com/chinglee-iot/FreeRTOS/tree/smp-integration-coverity-test).
2. cd FreeRTOS/Test/Coverity
3. Update the compiler configuration in Coverity
`cov-configure --force --compiler cc --comptype gcc -- -std=gnu90`
4. **Build the (pseudo) application ( different make target )**
`cov-build --emit-complementary-info --dir cov-out make singlecore`
5. Go to the Coverity output directory (cov-out) and begin Coverity static analysis
`cd cov-out/`
`cov-analyze --dir . --coding-standard-config ../coverity_misra.config`
6. Generate the report with the following command
`cov-format-errors --dir . --file "Source/tasks.c" --exclude-files '(/Test/|/Source/portable/)' --html-output html-out;`

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by ActoryOu — APPROVED** [review]

File: tasks.c [hunk 1]
```diff
@@ -391,6 +391,9 @@ typedef tskTCB TCB_t;
 #if ( configNUMBER_OF_CORES == 1 )
     portDONT_DISCARD PRIVILEGED_DATA TCB_t * volatile pxCurrentTCB = NULL;
 #else
+/* MISRA Ref 8.4.1 [Declaration shall be visible] */
+/* More details at: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/main/MISRA.md#rule-84 */
+/* coverity[misra_c_2012_rule_8_4_violation] */
```
Comment (chinglee-iot) [inline tasks.c]: Port use this variable in context swtich.

File: tasks.c [hunk 2]
```diff
@@ -937,6 +951,9 @@ static void prvAddNewTaskToReadyList( TCB_t * pxNewTCB ) PRIVILEGED_FUNCTION;
             if( listLIST_IS_EMPTY( &( pxReadyTasksLists[ uxCurrentPriority ] ) ) == pdFALSE )
             {
                 List_t * const pxReadyList = &( pxReadyTasksLists[ uxCurrentPriority ] );
+                /* MISRA Ref 11.3.2 [Cast to different type] */
+                /* More details at: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/main/MISRA.md#rule-113 */
+                /* coverity[misra_c_2012_rule_11_3_violation] */
```
Comment (chinglee-iot) [inline tasks.c]: minimal list item.

File: tasks.c [hunk 3]
```diff
@@ -1166,8 +1186,11 @@ static void prvAddNewTaskToReadyList( TCB_t * pxNewTCB ) PRIVILEGED_FUNCTION;
         {
             /* The memory used for the task's TCB and stack are passed into this
              * function - use them. */
+            /* MISRA Ref 11.3.1 [Cast to different type] */
+            /* More details at: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/main/MISRA.md#rule-113 */
+            /* coverity[misra_c_2012_rule_11_3_violation] */
```
Comment (chinglee-iot) [inline tasks.c]: Cast a buffer to TCB_t structure.

File: tasks.c [hunk 4]
```diff
@@ -1225,8 +1248,11 @@ static void prvAddNewTaskToReadyList( TCB_t * pxNewTCB ) PRIVILEGED_FUNCTION;
             /* Allocate space for the TCB.  Where the memory comes from depends
              * on the implementation of the port malloc function and whether or
              * not static allocation is being used. */
+            /* MISRA Ref 11.3.1 [Cast to different type] */
+            /* More details at: https://github.com/FreeRTOS/FreeRTOS-Kernel/blob/main/MISRA.md#rule-113 */
+            /* coverity[misra_c_2012_rule_11_3_violation] */
```
Comment (chinglee-iot) [inline tasks.c]: Cast a buffer to TCB_t

File: MISRA.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+#### Rule 8.6
+
+_Ref 8.6.1_
+
+- MISRA C:2012 Rule 8.6: An identifier with external linkage shall have exactly one external definition.
+        Function is declared in header file correctly. But the function name in task.h/tasks.c is different if portUSING_MPU_WRAPPERS enabled.
+        
+#### Rule 8.9
+
+_Ref 8.9.1_
```
Comment (chinglee-iot) [inline MISRA.md]: Remove this

File: MISRA.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        False alarm. This variable is actually used by several different functions.
+        
+#### Rule 11.3
+
+_Ref 11.3.1_
+
+- MISRA C:2012 Rule 11.3: A cast shall not be performed between a pointer to object type and a pointer to a different object type.
+        Unusual cast is ok as the structures are designed to have the same alignment, this is checked by an assert.
+
+_Ref 11.3.2_
```
Comment (chinglee-iot) [inline MISRA.md]: Turn of mini item list

File: tasks.c [hunk 7]
```diff
@@ -946,6 +963,9 @@ static void prvAddNewTaskToReadyList( TCB_t * pxNewTCB ) PRIVILEGED_FUNCTION;
 
                 for( pxIterator = listGET_HEAD_ENTRY( pxReadyList ); pxIterator != pxEndMarker; pxIterator = listGET_NEXT( pxIterator ) )
                 {
+                    /* MISRA Ref 11.5.1 [Cast to different type] */
```
Comment (chinglee-iot) [inline tasks.c]: Use cast instead of suppression.

File: tasks.c [hunk 8]
```diff
@@ -1288,11 +1314,14 @@ static void prvAddNewTaskToReadyList( TCB_t * pxNewTCB ) PRIVILEGED_FUNCTION;
             /* Allocate space for the TCB.  Where the memory comes from depends
              * on the implementation of the port malloc function and whether or
              * not static allocation is being used. */
+            /* MISRA Ref 11.5.1 [Cast to different type] */
```
Comment (chinglee-iot) [inline tasks.c]: Remove suppression here.

File: MISRA.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+- MISRA C:2012 Rule 11.3: A cast shall not be performed between a pointer to object type and a pointer to a different object type.
+        The mini list structure is used as the list end to save RAM.  This is checked and valid.
+  
+#### Rule 11.5
+
+_Ref 11.5.1_
```
Comment (chinglee-iot) [inline MISRA.md]: Remove this

File: tasks.c [hunk 10]
```diff
@@ -3111,6 +3140,9 @@ static BaseType_t prvCreateIdleTasks( void )
 
             for( x = ( BaseType_t ) 0; x < ( BaseType_t ) configMAX_TASK_NAME_LEN; x++ )
             {
+                /* MISRA Ref 18.1.1 [Overrun] */
```
Comment (chinglee-iot) [inline tasks.c]: Remove this one.

File: MISRA.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+- MISRA C:2012 Rule 11.5: A conversion should not be performed from pointer to void into pointer to object.
+        void * is used as this macro is used with timers and co-routines too.  Alignment is known to be fine as the type of the pointer stored and retrieved is the same.
+        
+#### Rule 18.1
```
Comment (chinglee-iot) [inline MISRA.md]: Remove

File: tasks.c (hunk 1 above)
Comment (chinglee-iot) [reply tasks.c]: Remove the supression here.

**Review by aggarg — APPROVED** [review]

**Pull request merged**